crypto/sparse_array/sparse_array.c

```diff
@@ -0,0 +1,180 @@
+#include "internal/sparse_array.h"
+
+#ifdef OPENSSL_SMALL_FOOTPRINT
+#define SA_BLOCK_BITS           4
+#else
+#define SA_BLOCK_BITS           12
+#endif
+#define SA_BLOCK_MAX            (1 << SA_BLOCK_BITS)
+#define SA_BLOCK_MASK           (SA_BLOCK_MAX - 1)
+#define SA_BLOCK_MAX_LEVELS     (((int)sizeof(size_t) * 8 + SA_BLOCK_BITS - 1) \
+                                 / SA_BLOCK_BITS)
+
+struct sparse_array_st {
+    int levels;
+    size_t top;
+    size_t nelem;
+    void **nodes;
+};
+
+OPENSSL_SA *OPENSSL_SA_new(void)
+{
+    OPENSSL_SA *res = OPENSSL_zalloc(sizeof(*res));
+
+    return res;
+}
+
+static void sa_doall(OPENSSL_SA *sa, void (*node)(void **),
+                     void (*leaf)(void *, void *), void *arg)
+{
+    int i[SA_BLOCK_MAX_LEVELS];
+    void *nodes[SA_BLOCK_MAX_LEVELS];
+    int l = 0;
+
+    i[0] = 0;
+    nodes[0] = sa->nodes;
+    while (l >= 0) {
+        const int n = i[l];
+        void ** const p = nodes[l];
+
+        if (n >= SA_BLOCK_MAX) {
+            if (p != NULL && node != NULL)
+                (*node)(p);
+            l--;
+        } else {
+            i[l] = n + 1;
+            if (p != NULL && p[n] != NULL) {
+                if (l < sa->levels - 1) {
+                    i[++l] = 0;
+                    nodes[l] = p[n];
+                } else if (leaf != NULL) {
+                    (*leaf)(p[n], arg);
+                }
+            }
+        }
+    }
+}
+
+static void sa_free_node(void **p)
+{
+    OPENSSL_free(p);
+}
+
+static void sa_free_leaf(void *p, void *arg)
+{
+    OPENSSL_free(p);
+}
+
+void OPENSSL_SA_free(OPENSSL_SA *sa)
+{
+    sa_doall(sa, &sa_free_node, NULL, NULL);
+    OPENSSL_free(sa);
+}
+
+void OPENSSL_SA_free_leaves(OPENSSL_SA *sa)
+{
+    sa_doall(sa, &sa_free_node, &sa_free_leaf, NULL);
+    OPENSSL_free(sa);
+}
+
+/* Wrap this in a structure to avoid compiler warnings */
+struct trampoline_st {
+    void (*func)(void *);
+};
+
+static void trampoline(void *l, void *arg)
+{
+    ((const struct trampoline_st *)arg)->func(l);
+}
+
+void OPENSSL_SA_doall(OPENSSL_SA *sa, void (*leaf)(void *))
+{
+    struct trampoline_st tramp;
+
+    tramp.func = leaf;
+    if (sa != NULL)
+        sa_doall(sa, NULL, &trampoline, &tramp);
+}
+
+void OPENSSL_SA_doall_arg(OPENSSL_SA *sa, void (*leaf)(void *, void *), void *arg)
+{
+    if (sa != NULL)
+        sa_doall(sa, NULL, leaf, arg);
+}
+
+size_t OPENSSL_SA_num(const OPENSSL_SA *sa)
+{
+    return sa == NULL ? 0 : sa->nelem;
+}
+
+void *OPENSSL_SA_get(OPENSSL_SA *sa, size_t n)
+{
+    int level;
+    void **p, *r = NULL;
+
+    if (sa == NULL)
+        return NULL;
+
+    if (n <= sa->top) {
+        p = sa->nodes;
+        for (level = sa->levels - 1; p != NULL && level > 0; level--)
+            p = (void **)p[(n >> (SA_BLOCK_BITS * level)) & SA_BLOCK_MASK];
+        r = p == NULL ? NULL : p[n & SA_BLOCK_MASK];
+    }
+    return r;
+}
+
+static void **alloc_node(void)
+{
+    return calloc(SA_BLOCK_MAX, sizeof(void *));
```

Comment:
Should we be using one of the `OPENSSL_` prefixed malloc functions?

Reply:
Yes :(